applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp

```diff
@@ -0,0 +1,754 @@
+
+    //Resetting the RHS
+    if ( rRightHandSideVector.size() != N_DOF )
+        rRightHandSideVector.resize( N_DOF, false );
+    noalias( rRightHandSideVector ) = ZeroVector( N_DOF );
+
+    //calculation flags
+    const bool CalculateStiffnessMatrixFlag = true;
+    const bool CalculateResidualVectorFlag = true;
+
+    CalculateAll(rLeftHandSideMatrix,
+                 rRightHandSideVector,
+                 rCurrentProcessInfo,
+                 CalculateStiffnessMatrixFlag,
+                 CalculateResidualVectorFlag);
+
+    KRATOS_CATCH( "" )
+}
+
+//----------------------------------------------------------------------------------------
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwBaseElement<TDim,TNumNodes>::
+    CalculateLeftHandSide( MatrixType& rLeftHandSideMatrix,
+                           const ProcessInfo& rCurrentProcessInfo )
+{
+    KRATOS_TRY;
+
+    // Calculation flags
+    const bool CalculateStiffnessMatrixFlag = true;
+    const bool CalculateResidualVectorFlag = false;
+    VectorType TempVector;
+
+    CalculateAll(rLeftHandSideMatrix,
+                 TempVector,
+                 rCurrentProcessInfo,
+                 CalculateStiffnessMatrixFlag,
+                 CalculateResidualVectorFlag);
+
+    KRATOS_CATCH("");
+}
+
+//----------------------------------------------------------------------------------------
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwBaseElement<TDim,TNumNodes>::
+    CalculateRightHandSide( VectorType& rRightHandSideVector,
+                            const ProcessInfo& rCurrentProcessInfo )
+{
+    KRATOS_TRY
+
+    const unsigned int N_DOF = TNumNodes * (TDim + 1);
+
+    //Resetting the RHS
+    if ( rRightHandSideVector.size() != N_DOF )
+        rRightHandSideVector.resize( N_DOF, false );
+    noalias( rRightHandSideVector ) = ZeroVector( N_DOF );
+
+    const bool CalculateStiffnessMatrixFlag = false;
+    const bool CalculateResidualVectorFlag = true;
+    MatrixType TempMatrix = Matrix();
+
+    CalculateAll(TempMatrix,
+                 rRightHandSideVector,
+                 rCurrentProcessInfo,
+                 CalculateStiffnessMatrixFlag,
+                 CalculateResidualVectorFlag);
+
+    KRATOS_CATCH( "" )
+}
+
+//----------------------------------------------------------------------------------------
+template< unsigned int TDim, unsigned int TNumNodes >
+void UPwBaseElement<TDim,TNumNodes>::
+    EquationIdVector(EquationIdVectorType& rResult,
+                     const ProcessInfo& rCurrentProcessInfo) const
+{
+    KRATOS_TRY
+
+    const GeometryType& rGeom = this->GetGeometry();
+    const unsigned int N_DOF = TNumNodes * (TDim + 1);
+    unsigned int index = 0;
+
+    if (rResult.size() != N_DOF)
+      rResult.resize( N_DOF, false );
+
+    if (TDim == 2)
+    {
+        for (unsigned int i = 0; i < TNumNodes; i++)
+        {
+            rResult[index++] = rGeom[i].GetDof(DISPLACEMENT_X).EquationId();
```

Comment:
just to tell that you could slightly optimize this access by getting the position of the variable in the database. see e.g.

https://github.com/KratosMultiphysics/Kratos/blob/master/applications/FluidDynamicsApplication/custom_elements/fluid_element.cpp#L288

anyway, it is only a minor optimization